### ISSUE TITLE
Remove bootloader read-write filesystem requirement

### ISSUE THREAD

**Issue by paradigm** [bedrocklinux/bedrocklinux-userland#67]
brn could at run time determine which mount points need to be altered for brs/preenable and remount those as read-write before calling brs, then remount them back to read-only to resume prior state before handing things off to the specified init.

**Comment by jawbonegroove** [comment]
perhaps allow some scripting here to handle a number of user-defined early rw tasks?

**Comment by paradigm** [comment]
Do you have any examples of what these early rw tasks would be?  Outside of setting up the chosen init system, I'm hard pressed to think of anything that could not be done equally well in an `/etc/rc.local` run by the chosen init system.  I could be missing something, though.  There's been a *lot* of confusion about `/bedrock/etc/fstab` vs `/etc/fstab`, and I'm concerned adding a `/bedrock/etc/rc.local` would further add to the confusion.  However, if there is a workflow which strictly requires it that I'm failing to consider there's no technical reason we couldn't add such a thing.

In fact, maybe I should pursue dropping `/bedrock/etc/fstab`.  Maybe a per-stratum config item would be less confusing for people.  There could be a configuration option which specifies the stratum's root filesystem and mount point.  `brs` could (re)mount that as appropriate when the stratum is enabled.  That may both resolve the `/bedrock/etc/fstab` vs `/etc/fstab` confusion *and* resolve the read-write filesystem issue.

**Comment by jawbonegroove** [comment]
I'd like to be able to set some things per-boot for debugging and general tinkering...maybe parse bootloader info and strata init mtimes to determine which stratum I want handling the init...multiple architecture bedrock bootable usb sticks, etc. I want to take full advantage of bedrock on arm, which means uboot info dumps and constant rebooting with new configs/kernels; it'd be handy to be able to automate that as much as possible /if needed/

I'm also looking at porting `redox-os` to arm/aarm64, and this is the sort of development niche where I think bedrock will really shine.

**Comment by paradigm** [comment]
Interesting.

> I'd like to be able to set some things per-boot for debugging and general tinkering...

If you put "debug_brn" on the kernel line in the bootloader, it'll drop you to a shell at that pre-boot time, as documented [here](http://bedrocklinux.org/1.0beta2/commands.html#brn).  That's not quite what you're asking for, but it may help a bit right now.  It kicks in immediately before running the selected init, after all the setup.

> maybe parse bootloader info and strata init mtimes to determine which stratum I want handling the init

From your last post I was thinking a simple /bedrock/etc/rc.local could be called somewhere during the very early boot, but if I follow that correctly you'd also need hooks into the system for selecting which init to use.  Unless you want to `sed` and `mv` over `brn.conf`'s default option, which feels icky to me.

One option is to rework `brn` such that it calls out to some external bit when it comes time to chose which init.  By default, that external bit would be something along the lines of the current menu, but it'd be easily adjusted to something which parses the bootloader info and looks at init mtimes.

My main concern here is that, at the moment, I can't imagine many other people using this, but I do expect people to be confused by yet another option.  In the past I've found Bedrock Linux can be *too* flexible in a way people find overwhelming.  If I reworked `brn` to select which init to use in a single, simple function call that you could easily comment out and replace with your own, would that be sufficient for your needs?  It's just a shell script - no recompiling or any such thing.  If I'm off point and you expect other people to use this I'm definitely not firm on it - I don't want to accidentally discard people here.

> I want to take full advantage of bedrock on arm

Nice!  [I know Bedrock Linux works on ARM](https://www.youtube.com/watch?v=O9dh2OnKjAU), but it's not well tested there.  If you find Bedrock Linux needs to make some issues to better support ARM do let me know.

> which means uboot info dumps and constant rebooting with new configs/kernels; it'd be handy to be able to automate that as much as possible /if needed/

If you're going to reboot anyways, does this automation need to happen before the init runs?  I'd imagine `/etc/rc.local` would suffice for it as well, no?

> I'm also looking at porting redox-os to arm/aarm64, and this is the sort of development niche where I think bedrock will really shine.

Oh, very nice!  I'm rather fond of Rust.  In Bedrock Linux's history there's been *one* issue that causes a Bedrock Linux component to crash - an issue that couldn't have happened under Rust.

I could definitely see Bedrock Linux aiding in development of projects such as Redox.

While you *probably* already know this if you're pursuing something like porting redox-os to arm/aarm64, I figure I should point out *just in case* that that Bedrock Linux is dependent on *specifically* Linux kernel features and quirks - I'd be very surprised to find it works with Redox's kernel.  The very nature of what Bedrock Linux is doing requires non-portable kernel nuance specific (ab)use.  Developing Redox OS from Linux should work fine, though.

-----

After thinking about it a fair bit, I'm concerned making a sufficiently flexible system that could meet all the expected workflows here would be add a lot of code and complexity, and that a less flexible system would be bordering on useless.  Maybe we can solve your goal here without the code base complexity cost...  It's worth noting that `brn` is a shell script, and at the moment I don't have any plans of moving away from that.  It should be pretty easy to add code in between the steps yourself if you're comfortable with even basic shell scripting.  If I rewrite `brn` so the bulk is abstracted away in functions and the "main" root-level code is just a conditional-free linear list of function calls such that it's trivial to add in code between them, would that suffice for your needs?  I'm strongly considering rewriting brn anyways, as there's quite a few changes that need to be made.

If that's not sufficient, I think what you're going for here should be a separate issue from this one.  Feel free to make a new issue which goes into a bit more depth for what you're looking for.  Things like:

- When you'd want the hooks to run (look [here](https://github.com/bedrocklinux/bedrocklinux-userland/blob/8c1345fa86988e2d7058d5778e91d04beb4531f2/src/slash-bedrock/sbin/brn#L445-L527)) to get a feel for the broad steps that are taken between which a hook could be potentially placed)
- What information needs to be made available to the hooks (e.g. if the hook is after the user selects the init, does it need access to what init was chosen?)
- What parts of `brn` the hooks should be able to change (e.g. indicating the desired init, indicating the init-choosing menu should not be presented, indicating which strata should be enabled before init is run, et al)
- Example use cases

**Comment by jawbonegroove** [comment]
actually i think 'debug_brn` is what i'm going to need first anyway, thanks!

    ...which feels icky to me.
following so correctly that you're leading. feels icky to me too.

    My main concern here is that, at the moment, I can't imagine many other people using this, but I do expect people to be confused by yet another option. In the past I've found Bedrock Linux can be too flexible in a way people find overwhelming
agreed. 

    `/etc/rc.local`
if i get it working right, i'll be using the same bedrock image/sdcard to boot several devices and will have the script assigning the correct device init, including `rc.local`

    ...couldn't have happened under Rust
interesting. i'm just getting into it but it's exciting to see a new os come from fresh ground, eh? it sure seems elegant

    ... nuance specific (ab)use. Developing Redox OS from Linux should work fine, though.
yes, i want bedrock (ideally the same image) running across a couple of allwinner and broadcomm chipsets, mostly to play with your toy (thanks btw, it's great ^^); but also, basically all arm distros are finicky and i think bedrock will do a lot to stabilise/globalise my rust/redox arm tinkerings from linux.

   ... making a sufficiently flexible system that could meet all the expected workflows here would be add a lot of code and complexity, and that a less flexible system would be bordering on useless ...
yeh we're totally on the same page. just some script space should suffice, so if you segment `... `brn` so the bulk is abstracted away in functions ...` then i'll be happy as a clam. for my own purposes, more than that seems like overkill and less the point, eh.

**Comment by paradigm** [comment]
> if i get it working right, i'll be using the same bedrock image/sdcard to boot several devices and will have the script assigning the correct device init, including rc.local

Choosing the correct init can't be done in `rc.local`, so sure that bit has to be in `brn`.

The phrase "init, including rc.local" makes me think you *might* be missing a useful Bedrock Linux nuance here.  [By default](https://github.com/bedrocklinux/bedrocklinux-userland/blob/8c1345fa86988e2d7058d5778e91d04beb4531f2/src/slash-bedrock/etc/frameworks.d/default#L39) `rc.local` is a [global file](http://bedrocklinux.org/1.0beta2/concepts.html#local-vs-global).  This means irrelevant of which stratum is providing init, they'll all see the same `/etc/rc.local` file.  Provided all the inits you are considering here respect that file (I've found some systemd distros do not do so by default - I may have Bedrock Linux detect this and add systemd configuration to enable it in the upcoming release), whatever you have in `/etc/rc.local` should be run by the selected init every time you boot even if you switch which init that is.  I regularly switch which init I'm using as part of testing/developing Bedrock, and so I put a lot of init-agnostic boot-time things in `/etc/rc.local`.

I may need to document/emphasize this better.

> thanks btw, it's great ^^

You are absolutely welcome!

> yeh we're totally on the same page. just some script space should suffice, so if you segment `... brn so the bulk is abstracted away in functions ...` then i'll be happy as a clam. for my own purposes, more than that seems like overkill and less the point, eh.

Sounds good.  While it's *probably* good enough now for you to utilize, I'll make an explicit point of moving all the conditional branching out of the main/root and into functions for the upcoming release.

**Comment by jawbonegroove** [comment]
> init-agnostic boot-time things in /etc/rc.local

this is the plan. I have not put tons of consideration into nuance here though; I was vaguely thinking source additional device-specific `rc.local` mush from a file copied from the correct strata during `brn`. 

I had to shelf some of my projects and will have to spend some time refamiliarizing myself. I have yet to set up a nyla box.  :x

> moving all the conditional branching out of the main/root and into functions

sounds brilliant to me! and for now i'll be alright with the debug shell.

**Comment by jawbonegroove** [comment]
side note, if anyone's playing with android on bedrock, this is their key to happiness and joy.

**Comment by paradigm** [comment]
> this is the plan. I have not put tons of consideration into nuance here though; I was vaguely thinking source additional device-specific rc.local mush from a file copied from the correct strata during brn. 

The copy seems unnecessary to me.  I'd be inclined to try and detect which device you're on and run the code in place.  If there is a one-to-one relationship between int-providing stratum and device, you could utilize the aliased explicit path, and do something like `source /bedrock/strata/init/etc/device.rc`.  `/bedrock/strata/init` will automatically redirect accordingly.

> side note, if anyone's playing with android on bedrock, this is their key to happiness and joy.

We had someone doing initial investigatory work on that, but sadly he moved away from the project.

**Comment by jawbonegroove** [comment]
._. will alised paths work across partitions? specifically, on arm I'm testing with 5 devices; for this "master sd card" idea i could probably make use of 9 inits between them. heh. 

yeah having an easy toe-in-the-door before android init will make a lot easier to tweak and take over .

**Comment by jawbonegroove** [comment]
wait that didn't come out right. I'll revisit this when I've slept.

**Comment by paradigm** [comment]
> ._. can alised paths be nested?

I don't think I understand the question.

Some background that may help to make sure we're on the same page:

For Bedrock to work, you'll end up with multiple instances of some file paths.  For example, both Debian and Ubuntu will expect different files at `/etc/apt/sources.list`, and so you end up with two `/etc/apt/sources.list` files.  Thus, a file path is no longer a unique identifier for a file.  "Strata" can be thought of as metadata which, in addition to a file path, uniquely identify a file.  So you could have a (debian, `/etc/apt/sources.list`) and a (ubuntu, `/etc/apt/sources.list`).

If you *just* specify the filename/path Bedrock Linux is pretty good at figuring out which one you *probably* want for everything to just work.  For example, if it sees Debian's `apt-get` reading a file at the path `/etc/apt/sources.list` it'll know to give it (debian, `/etc/apt/sources.list`).  If you run `reboot`, it'll know to automatically use the `reboot` associated with the currently in use init system.  If you run `man vim` from a `bash` prompt, where `bash` is provided by distro A, `man` is provided only by distro B, and vim's man page is provided only by distro C, Bedrock will figure the situation out and make it all work.  This system to figure out what comes from where is configurable - if you always want one distro to provide a given command even if others could provide it, you can express that to Bedrock Linux and it should just work from that point onward.

If Bedrock can't figure out from context which you want (can't read your mind), you'll have to specify it explicitly:

- For read or write access, you can use `/bedrock/strata/<stratum>/path/to/file`.  For example, to read specifically Debian's `/etc/apt/sources.list` you can read the file at `/bedrock/strata/debian/etc/apt/sources.list`.  If you want to read specifically Ubuntu's `/etc/apt/sources.list`, you can read the file at `/bedrock/strata/ubuntu/etc/apt/sources.list`.
- For execute aess, you use `brc <stratum> <cmd> <args>`.  So `brc debian apt-get install foo` will use Debian's `apt-get` to install `foo`, and `brc ubuntu apt-get install bar` will use Ubuntu's `apt-get` to install bar.

Additionally, Bedrock has aliases for strata.  The most notably one here is the init alias, which always redirects to the strata currently providing init.  If you reboot and select another init, the init alias will change accordingly.  `brc init` and `/bedrock/strata/init` both redirect accordingly.  `bri -a <alias>` will indicate where an alias is currently redirecting; for me right now, `bri -a init` kicks out `void` since I selected `void` to provide init when I last rebooted.  You may also have aliases for `local`, `rootfs` and `global` by default.  Some users do things like set up `debstable` to alias to the current stable Debian release (so at the moment, `debstable` -> `jessie`, but when Debian Stretch becomes stable they'll update it to `debstable` -> `stretch` then make an `deboldstable` point to `jessie`).

Give this, I don't quite follow how nesting makes sense.  I guess you could have one alias point to another which points to another - that should work.

The above is spelled out in more depth [here](http://bedrocklinux.org/1.0beta2/concepts.html).

> and work across partitions?

Yes, this whole system works across partitions.  However, complex partition schemes require more complicated Bedrock Linux configuration.  I'm hoping to eventually make it smarter about automatically detecting what it should do to minimize needed configuration, but we're not there yet.  You'll have to configure `/etc/fstab`, possibly the `bind` or `share` items in `strata.conf`, and maybe `/bedrock/etc/fstab`.  [Configuration documentation is here](http://bedrocklinux.org/1.0beta2/configure.html).

> specifically, on arm I'm testing with 5 devices; for this "master sd card" idea i could probably make use of 9 inits between them. heh. 

That should be fine.  I'm pretty sure I swapped between over a dozen during final testing for the current Bedrock Linux release.

> yeah having an easy toe-in-the-door before android init will make a lot easier to tweak and take over .

I assumed, perhaps incorrectly, that you were using "traditional" Linux distros on ARM, like Debian or Gentoo.  You're certainly welcome to try Android, but I'd recommend getting avoiding getting your hopes up about much success with Android + Bedrock, at least not without a substantial amount of work.  Amongst other things, Android does some really funky stuff with UID/GIDs, uses a completely different subsystem for determining which executable to run, and changes the file layout far from what Bedrock Linux's default configuration is aimed at.

**Comment by jawbonegroove** [comment]
> I don't think I understand the question.

I don't think so either but you answered it anyway. lol ...I've been awake a long time, sorry. I used hawky but ...maybe 2-3 years ago. I forget a lot.

**Comment by paradigm** [comment]
Bootloader read-write filesystem requirement removed in Poki.
